src/signature.c

```diff
@@ -24,198 +24,354 @@
+	if (!cert) {
+		fclose(fp_pubkey);
+		goto error;
 	}
-	if (crl_filename) {
-		X509_LOOKUP *lookup = X509_STORE_add_lookup(store, X509_LOOKUP_file());
-		if (!lookup) {
-			goto err;
-		}
-		if (X509_load_crl_file(lookup, crl_filename, X509_FILETYPE_PEM) != 1) {
-			goto err;
-		}
-		X509_STORE_set_flags(store, X509_V_FLAG_CRL_CHECK | X509_V_FLAG_CRL_CHECK_ALL);
+
+	pkey = X509_get_pubkey(cert);
+	if (!pkey) {
+		fclose(fp_pubkey);
+		X509_free(cert);
+		goto error;
 	}
-	return store;
-err:
-	X509_STORE_free(x509_store);
-	return NULL;
+	return true;
+error:
+	ERR_print_errors_fp(stderr);
+	return false;
 }
 
-bool signature_download_and_verify(const char *data_url, const char *data_filename)
+/* This is the main part of the signature validation.
+ * This function reads a file in chunks of 4096 bytes to create a hash from
+ * the content, then verifies the hash using the publick key against the
+ * signature.
+ *
+ * returns: true if signature was correct, false otherwise */
+static bool validate_signature(FILE *fp_data, FILE *fp_sig)
 {
-	char *sig_url;
-	char *sig_filename;
-	int ret;
-	bool result;
+	unsigned char *data;
+	unsigned char *signature;
+	size_t sig_len;
+	size_t data_size;
+	size_t count;
+	struct stat st;
+	int fd;
+	EVP_MD_CTX *ctx;
+
+	/* Read in the .sig file */
+	fd = fileno(fp_sig);
+	if (fstat(fd, &st) != 0) {
+		fprintf(stderr, "Failed to stat .sig file\n");
+	}
+	sig_len = st.st_size;
+	signature = malloc(sig_len);
+	count = fread(signature, 1, sig_len, fp_sig);
+	if (count != sig_len) {
+		fprintf(stderr, "Failed to get signature length! Read %lu/%lu bytes\n", count, sig_len);
+		goto error;
+	}
+
+	/* Read in the corresponding data file for the .sig */
+	fd = fileno(fp_data);
+	if (fstat(fd, &st) != 0) {
+		fprintf(stderr, "Failed to stat data file\n");
+		goto error;
+	}
+	data_size = st.st_size;
+	data = malloc(data_size);
+	count = fread(data, 1, data_size, fp_data);
+	if (count != data_size) {
+		fprintf(stderr, "Failed to read full data file\n");
+		goto error;
+	}
 
-	string_or_die(&sig_url, "%s.signed", data_url);
+	ctx = EVP_MD_CTX_create();
+	/* Initialize verify context and use sha256 algorithm internally */
+	if (!EVP_VerifyInit_ex(ctx, EVP_sha256(), NULL)) {
+		goto error;
+	}
 
-	string_or_die(&sig_filename, "%s.signed", data_filename);
+	/* Hashes data_size bytes into the verification context */
+	if (!EVP_VerifyUpdate(ctx, data, data_size)) {
+		goto error;
+	}
+	/* Verify data in context using pkey against signature */
+	if (EVP_VerifyFinal(ctx, signature, sig_len, pkey) == 1) {
+		EVP_MD_CTX_destroy(ctx);
+		free(data);
+		free(signature);
+		return true;
+	}
 
-	ret = swupd_curl_get_file(sig_url, sig_filename, NULL, NULL, false);
-	if (ret) {
-		result = false;
-	} else {
-		result = signature_verify(data_filename, sig_filename);
+error:
+	if (signature) {
+		free(signature);
 	}
-	if (!result) {
-		unlink(sig_filename);
+	if (data) {
+		free(data);
```

Comment:
I'm seeing a compiler warning:

```
src/signature.c:237:3: warning: ‘data’ may be used uninitialized in this function [-Wmaybe-uninitialized]
```

That pointer should be initialized to NULL.